### PR TITLE
Introduce `wcpay_payment_request_is_cart_supported` filter.

### LINKED ISSUE #7438: Introduce product based filter to `WC_Payments_Payment_Request_Button_Handler::has_allowed_items_in_cart()`
### Description

Introduce a filter to allow developers to modify whether a product allows the express pay options (Google and Apple Pay) in the cart.

```php
/**
 * Filter whether product supports Payment Request Button on cart page.
 *
 * @param boolean $is_supported Whether product supports Payment Request Button on cart page.
 * @param object  $_product     Product object.
 */
apply_filters( 'wcpay_payment_request_is_cart_supported', true, $_product )
```

This is intended as a sibling filter to the existing `wcpay_payment_request_is_product_supported` filter.

The purpose of the new filter is to allow developers to specify whether a product supports the express payment options based on attributes other than type.

An example is bookings products generally support the express pay options in the cart and at checkout, but for bookings requiring confirmation the express pay options are not supported as payment is not processed at the time.

Introducing a filter allows extenders to apply more complex rules without requiring compatibility shims in Payments.

**Possible alternative approaches:**

* Use the existing `woocommerce_cart_item_product` by allowing developers to return a non-object to indicate the product is not supported (ideally the developer would return false) -- this would require changes to the has_subscriptions method
* pass the product as a parameter to the filter `wcpay_payment_request_supported_types`

### Acceptance criteria

* [Supported product types](https://github.com/peterwilsoncc/woocommerce-payments/blob/efd9c5a5a0068972eab6827967eac6c8be4ccf6e/includes/class-wc-payments-payment-request-button-handler.php#L552-L568) takes priority over the new filter.
* Filter only runs for supported product types

### Testing instructions

1. Enable express payment options (Google and Apple Pay)
   * testing Apple Pay requires you use the Safari browser
   * testing Google Pay requires you use the Chrome browser
2. Install WooCommerce Bookings
3. Create a bookable product that requires confirmation
4. Order the item by clicking "check availability" to place it in your cart
5. Click through to the cart
6. The express pay option should not be displayed
7. Click through to checkout
8. The express pay option should not be displayed
9. Empty the cart
10. Create a bookable product that does not require confirmation
11. Add the item to your cart
12. Click through to the cart
13. The express pay option should be displayed
14. Click through to checkout
15. The express pay option should be displayed

### Additional context

https://github.com/woocommerce/woocommerce-bookings/issues/3619

### PR DESCRIPTION
**Pull request by peterwilsoncc**
Fixes #7438

#### Changes proposed in this Pull Request

Introduces the filter `wcpay_payment_request_is_cart_supported` as a sibling filter to `wcpay_payment_request_is_product_supported`. 

This is to allow extenders to specify whether a product supports the express payment options based on attributes other than type.

An example is bookings products generally support the express pay options in the cart and at checkout, but for bookings requiring confirmation the express pay options are not supported as payment is not processed at the time.

Reviewers:

* Please advise if and how you would like this to be tested.

#### Testing instructions

1. Enable express payments options (Apple or Google Pay)
   * testing Apple Pay requires you use the Safari browser
   * testing Google Pay requires you use the Chrome browser
2. Create a simple product
3. Add the product to your cart
4. View the cart
5. The express pay button should display
6. Add the mini plugin `add_filter( 'wcpay_payment_request_is_cart_supported', '__return_false' )` to your mu-plugins folder
7. Refresh the cart, the express pay button should not display
8. Delete the mini-plugin to avoid confusion later

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by jeffpaul** [comment]
@jessy-p can you help slot this for review in hopes of unblocking the downstream work in https://github.com/woocommerce/woocommerce-bookings/issues/3619#issuecomment-1752271221 && https://github.com/woocommerce/woocommerce-bookings-availability/issues/545#issuecomment-1752309560?

File: changelog/dev-7438-payment-cart-supported-filter [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: dev
+
+Introduce filter `wcpay_payment_request_is_cart_supported`.
```
Comment (dpaun1985) [inline changelog/dev-7438-payment-cart-supported-filter]: could you elaborate a little on the purpose of this filter?

File: includes/class-wc-payments-payment-request-button-handler.php [hunk 2]
```diff
@@ -585,6 +585,18 @@ public function has_allowed_items_in_cart() {
 				return false;
 			}
 
+			/**
+			 * Filter whether product supports Payment Request Button on cart page.
+			 *
+			 * @since x.x.x
```
Comment (dpaun1985) [inline includes/class-wc-payments-payment-request-button-handler.php]: this needs to be updated.

**Review by dpaun1985 — COMMENTED** [review]
✅ LGTM
The code looks good to me. I also tested it and works as expected. 
However I have 2 small comments.

File: includes/class-wc-payments-payment-request-button-handler.php (hunk 2 above)
Comment (mordeth) [reply includes/class-wc-payments-payment-request-button-handler.php]: +1

**Review by mordeth — APPROVED** [review]
✅ Code looks good!
✅ Manually tested and works as expected!

File: changelog/dev-7438-payment-cart-supported-filter (hunk 1 above)
Comment (peterwilsoncc) [reply changelog/dev-7438-payment-cart-supported-filter]: Expanded the detail in fc5ee98a

File: includes/class-wc-payments-payment-request-button-handler.php (hunk 2 above)
Comment (peterwilsoncc) [reply includes/class-wc-payments-payment-request-button-handler.php]: @dpaun1985 What version should I change this to? I've checked the allow maintainers to edit button on the PR to allow folks to modify this once the PR is added to a milestone.

Comment (dpaun1985) [reply includes/class-wc-payments-payment-request-button-handler.php]: ok great. let's leave it like this for now. LGTM.

**Review by dpaun1985 — APPROVED** [review]

File: includes/class-wc-payments-payment-request-button-handler.php (hunk 2 above)
Comment (frosso) [reply includes/class-wc-payments-payment-request-button-handler.php]:
```suggestion
			 * @since 6.9.0
```

**Review by frosso — APPROVED** [review]

**Pull request merged**